src/libstore/derivations.cc

```diff
@@ -1097,36 +1097,52 @@ Derivation Derivation::fromJSON(
     const Store & store,
     const nlohmann::json & json)
 {
+    using nlohmann::detail::value_t;
+
     Derivation res;
 
-    res.name = json["name"];
+    ensureType(json, value_t::object);
 
-    {
-        auto & outputsObj = json["outputs"];
-        for (auto & [outputName, output] : outputsObj.items()) {
+    res.name = valueOfTypeAt(json, value_t::string, "name");
+
+    try {
+        auto & outputsObj = valueOfTypeAt(json, value_t::object, "outputs");
```

Comment:
But now you will get a double "while" trace item. I really think it is better to just use the two functions separately.

Reply:
That was intentional. My idea was to give the user as much information as possible so they can figure out where in the file the problem is. I also don't think double "while" traces are a problem themselves, multiple steps are kind of the idea of a trace, right?

But I've dug deeper and this is not a problem we should solve ourselves. Maybe in the future we could try setting [`JSON_DIAGNOSTICS`](https://json.nlohmann.me/api/macros/json_diagnostics/), which effectively gives a traceback where in the JSON the error happened, but that's a story for another time.

I followed your advice and removed this function.